File: backend/automate/Automate/AutoApp/services/gas_stations.py
from __future__ import annotations

from typing import Any, Dict, List

from AutoApp.models import Fueling, User, Car


def list_gas_station_cards(user: User, car: Car, limit: int = 50) -> List[Dict[str, Any]]:
    qs = (
        Fueling.objects.filter(user=user, car=car)
        .select_related("gas_station", "fuel_type")
        .order_by("-date")[:limit]
    )

    cards = []
    for f in qs:
        gs = f.gas_station
        cards.append({
            "fueling_id": f.fueling_id,
            "date": f.date,
            "price_per_liter": float(f.price_per_liter),
            "fuel_type": f.fuel_type.name,
            "gas_station": {
                "gas_station_id": gs.gas_station_id,
                "name": gs.name,
                "city": gs.city,
                "postal_code": gs.postal_code,
                "street": gs.street,
                "house_number": gs.house_number,
            } if gs else None,
        })
    return cards

File: frontend/src/actions/gasStations.js
function formatDate(iso) {
  if (!iso) return "-"
  const d = new Date(iso)
  if (Number.isNaN(d.getTime())) return "-"
  const y = d.getFullYear()
  const m = String(d.getMonth() + 1).padStart(2, "0")
  const day = String(d.getDate()).padStart(2, "0")
  return `${y}. ${m}. ${day}.`
}

function buildAddress(gs) {
  if (!gs) return "-"
  const parts = [gs.street, gs.house_number].filter(Boolean)
  if (parts.length > 0) return parts.join(" ")
  return gs.name || "-"
}

function handleUnauthorized(response) {
  if (response.status === 401) {
    localStorage.clear()
    window.location.href = "/"
    throw new Error("Lejárt a belépés. Jelentkezz be újra.")
  }
}

export async function getGasStations() {
  const token = localStorage.getItem("token")
  const carId = localStorage.getItem("selected_car_id")
  const url = carId
    ? `http://localhost:8000/api/gas-stations/?car_id=${carId}`
    : "http://localhost:8000/api/gas-stations/"

  const response = await fetch(url, {
    headers: { Authorization: `Bearer ${token}` },
  })

  const data = await response.json().catch(() => ({}))
  handleUnauthorized(response)
  if (!response.ok) {
    throw new Error(data.detail || "Nem sikerült lekérdezni a benzinkutakat")
  }

  const items = data.gas_station_cards ?? []
  return items.map((item) => ({
    id: item.fueling_id,
    datum: formatDate(item.date),
    helyseg: item.gas_station?.city || "-",
    cim: buildAddress(item.gas_station),
    literft: Number(item.price_per_liter || 0),
    fuelType: item.fuel_type || "-",
    gasStationId: item.gas_station?.gas_station_id || null,
  }))
}